File: backend/routes/scanners.py
from fastapi import APIRouter, HTTPException, File, UploadFile
from pydantic import BaseModel
from typing import Optional
import random
from supabase_client import supabase_admin
from config import BODY_SCANNER_API_KEY, FACE_SCANNER_API_KEY, FOOD_SCANNER_API_KEY

router = APIRouter(prefix="/api/scan", tags=["scanners"])

class ScanRequest(BaseModel):
    user_id: str
    scan_type: str

@router.post("/body")
async def scan_body(user_id: str, file: UploadFile = File(...)):
    try:
        xp_earned = 8

        analysis_result = {
            "posture": "Slight forward head tilt",
            "postureDetails": "Mild forward head posture",
            "composition": "Lean-moderate",
            "muscle": f"{random.randint(35, 45)}%",
            "bodyType": random.choice(["Mesomorph", "Ectomorph", "Endomorph"]),
            "recommendations": [
                "Maintain good posture throughout the day",
                "Include strength training 3-4x per week",
                "Focus on compound movements"
            ]
        }

        scan_data = {
            "user_id": user_id,
            "scan_type": "body",
            "analysis_result": analysis_result,
            "xp_earned": xp_earned
        }

        response = supabase_admin.table("scans").insert(scan_data).execute()

        user_response = supabase_admin.table("users").select("xp, level").eq("id", user_id).execute()
        if user_response.data:
            user = user_response.data[0]
            new_xp = user.get("xp", 0) + xp_earned
            current_level = user.get("level", 1)

            xp_for_next_level = current_level * 100
            if new_xp >= xp_for_next_level:
                current_level += 1
                new_xp -= xp_for_next_level

            supabase_admin.table("users").update({
                "xp": new_xp,
                "level": current_level
            }).eq("id", user_id).execute()

        return {
            "message": "Body scan completed successfully",
            "analysis": analysis_result,
            "xp_earned": xp_earned,
            "scan_id": response.data[0]["id"] if response.data else None
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Body scan failed: {str(e)}")

@router.post("/face")
async def scan_face(user_id: str, file: UploadFile = File(...)):
    try:
        xp_earned = 6

        analysis_result = {
            "skinType": random.choice(["Combination", "Oily", "Dry", "Normal"]),
            "description": "Balanced skin with minor concerns",
            "concerns": random.choice([
                "Minor acne, slight tone unevenness",
                "Dry patches on cheeks",
                "Oily T-zone",
                "Fine lines around eyes"
            ]),
            "aiSuggestion": "Gentle cleansing, niacinamide AM, vitamin C AM, retinol PM, SPF 50+",
            "recommendedProduct": random.choice([
                "Youth-Glow Serum",
                "Hydration Boost Cream",
                "Clear Skin Toner"
            ]),
            "glowScore": random.randint(65, 85)
        }

        scan_data = {
            "user_id": user_id,
            "scan_type": "face",
            "analysis_result": analysis_result,
            "xp_earned": xp_earned
        }

        response = supabase_admin.table("scans").insert(scan_data).execute()

        user_response = supabase_admin.table("users").select("xp, level").eq("id", user_id).execute()
        if user_response.data:
            user = user_response.data[0]
            new_xp = user.get("xp", 0) + xp_earned
            current_level = user.get("level", 1)

            xp_for_next_level = current_level * 100
            if new_xp >= xp_for_next_level:
                current_level += 1
                new_xp -= xp_for_next_level

            supabase_admin.table("users").update({
                "xp": new_xp,
                "level": current_level
            }).eq("id", user_id).execute()

        return {
            "message": "Face scan completed successfully",
            "analysis": analysis_result,
            "xp_earned": xp_earned,
            "scan_id": response.data[0]["id"] if response.data else None
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Face scan failed: {str(e)}")

@router.post("/food")
async def scan_food(user_id: str, file: UploadFile = File(...)):
    try:
        xp_earned = 5

        analysis_result = {
            "foodName": random.choice(["Grilled Chicken Salad", "Pasta Carbonara", "Salmon Bowl", "Veggie Wrap"]),
            "nutrition": {
                "calories": random.randint(400, 700),
                "protein": random.randint(20, 40),
                "carbs": random.randint(40, 80),
                "fat": random.randint(10, 25)
            },
            "suggestion": random.choice([
                "Add more greens",
                "Reduce portion size",
                "Good balanced meal",
                "Include more protein"
            ]),
            "recommendation": random.choice([
                "Include more fiber",
                "Add healthy fats",
                "Great choice!",
                "Consider whole grains"
            ])
        }

        scan_data = {
            "user_id": user_id,
            "scan_type": "food",
            "analysis_result": analysis_result,
            "xp_earned": xp_earned
        }

        response = supabase_admin.table("scans").insert(scan_data).execute()

        user_response = supabase_admin.table("users").select("xp, level").eq("id", user_id).execute()
        if user_response.data:
            user = user_response.data[0]
            new_xp = user.get("xp", 0) + xp_earned
            current_level = user.get("level", 1)

            xp_for_next_level = current_level * 100
            if new_xp >= xp_for_next_level:
                current_level += 1
                new_xp -= xp_for_next_level

            supabase_admin.table("users").update({
                "xp": new_xp,
                "level": current_level
            }).eq("id", user_id).execute()

        return {
            "message": "Food scan completed successfully",
            "analysis": analysis_result,
            "xp_earned": xp_earned,
            "scan_id": response.data[0]["id"] if response.data else None
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Food scan failed: {str(e)}")

@router.get("/{user_id}/scans")
async def get_user_scans(user_id: str, scan_type: Optional[str] = None):
    try:
        query = supabase_admin.table("scans").select("*").eq("user_id", user_id)

        if scan_type:
            query = query.eq("scan_type", scan_type)

        response = query.order("timestamp", desc=True).execute()

        return response.data
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get scans: {str(e)}")

File: backend/routes/scanners.js
import express from 'express';
import multer from 'multer';
import OpenAI from 'openai';
import supabase from '../lib/supabase.js';
import authenticateUser from '../middleware/auth.js';

const router = express.Router();
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY || 'dummy-key'
});

const storage = multer.memoryStorage();
const upload = multer({
  storage,
  limits: { fileSize: 10 * 1024 * 1024 }
});

const calculateXPReward = (scanType) => {
  const rewards = {
    food: 5,
    body: 8,
    face: 6
  };
  return rewards[scanType] || 5;
};

const updateUserXP = async (userId, xpToAdd) => {
  const { data: user, error: fetchError } = await supabase
    .from('users')
    .select('xp, level, streak_days, last_activity_date')
    .eq('id', userId)
    .single();

  if (fetchError) throw fetchError;

  const newXP = user.xp + xpToAdd;
  const currentLevel = user.level;
  const xpForNextLevel = currentLevel * 100;
  const newLevel = Math.floor(newXP / 100) + 1;

  const today = new Date().toISOString().split('T')[0];
  const lastActivity = user.last_activity_date;
  let newStreakDays = user.streak_days || 0;

  if (lastActivity) {
    const lastDate = new Date(lastActivity);
    const currentDate = new Date(today);
    const diffTime = currentDate - lastDate;
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));

    if (diffDays === 1) {
      newStreakDays += 1;
    } else if (diffDays > 1) {
      newStreakDays = 1;
    }
  } else {
    newStreakDays = 1;
  }

  const { error: updateError } = await supabase
    .from('users')
    .update({
      xp: newXP,
      level: newLevel,
      streak_days: newStreakDays,
      last_activity_date: today
    })
    .eq('id', userId);

  if (updateError) throw updateError;

  return { newXP, newLevel, leveledUp: newLevel > currentLevel };
};

router.post('/food', authenticateUser, upload.single('image'), async (req, res) => {
  try {
    const userId = req.user.id;
    const imageFile = req.file;

    if (!imageFile) {
      return res.status(400).json({ error: 'Image file is required' });
    }

    const base64Image = imageFile.buffer.toString('base64');
    const imageDataUrl = `data:${imageFile.mimetype};base64,${base64Image}`;

    let analysis = {
      food_items: [],
      total_calories: 0,
      macros: { protein: 0, carbs: 0, fats: 0 },
      health_score: 0,
      recommendations: []
    };

    if (process.env.OPENAI_API_KEY && process.env.OPENAI_API_KEY !== 'dummy-key') {
      try {
        const response = await openai.chat.completions.create({
          model: 'gpt-4o',
          messages: [
            {
              role: 'user',
              content: [
                {
                  type: 'text',
                  text: 'Analyze this food image and provide: 1) List of food items, 2) Estimated calories, 3) Macros (protein, carbs, fats in grams), 4) Health score (0-100), 5) Nutritional recommendations. Format as JSON.'
                },
                {
                  type: 'image_url',
                  image_url: { url: imageDataUrl }
                }
              ]
            }
          ],
          max_tokens: 500
        });

        const aiResponse = response.choices[0].message.content;
        try {
          analysis = JSON.parse(aiResponse);
        } catch {
          analysis.recommendations = [aiResponse];
        }
      } catch (aiError) {
        console.error('OpenAI API error:', aiError);
        analysis.recommendations = ['AI analysis unavailable. Using mock data.'];
      }
    } else {
      analysis = {
        food_items: ['Healthy meal'],
        total_calories: 450,
        macros: { protein: 25, carbs: 40, fats: 15 },
        health_score: 75,
        recommendations: ['Great balanced meal! Consider adding more vegetables.']
      };
    }

    const xpAwarded = calculateXPReward('food');

    const { data: scanResult, error: insertError } = await supabase
      .from('scanner_results')
      .insert({
        user_id: userId,
        scan_type: 'food',
        image_url: imageDataUrl,
        analysis_data: analysis,
        xp_awarded: xpAwarded
      })
      .select()
      .single();

    if (insertError) throw insertError;

    const xpUpdate = await updateUserXP(userId, xpAwarded);

    res.json({
      message: 'Food scan completed successfully',
      scan: scanResult,
      analysis,
      xp_awarded: xpAwarded,
      ...xpUpdate
    });
  } catch (error) {
    console.error('Food scan error:', error);
    res.status(500).json({ error: 'Food scan failed' });
  }
});

router.post('/body', authenticateUser, upload.single('image'), async (req, res) => {
  try {
    const userId = req.user.id;
    const imageFile = req.file;

    if (!imageFile) {
      return res.status(400).json({ error: 'Image file is required' });
    }

    const base64Image = imageFile.buffer.toString('base64');
    const imageDataUrl = `data:${imageFile.mimetype};base64,${base64Image}`;

    let analysis = {
      posture_score: 0,
      body_composition: {},
      measurements: {},
      recommendations: []
    };

    if (process.env.OPENAI_API_KEY && process.env.OPENAI_API_KEY !== 'dummy-key') {
      try {
        const response = await openai.chat.completions.create({
          model: 'gpt-4o',
          messages: [
            {
              role: 'user',
              content: [
                {
                  type: 'text',
                  text: 'Analyze this body posture image and provide: 1) Posture score (0-100), 2) Body composition insights, 3) Estimated measurements, 4) Fitness recommendations. Format as JSON.'
                },
                {
                  type: 'image_url',
                  image_url: { url: imageDataUrl }
                }
              ]
            }
          ],
          max_tokens: 500
        });

        const aiResponse = response.choices[0].message.content;
        try {
          analysis = JSON.parse(aiResponse);
        } catch {
          analysis.recommendations = [aiResponse];
        }
      } catch (aiError) {
        console.error('OpenAI API error:', aiError);
        analysis.recommendations = ['AI analysis unavailable. Using mock data.'];
      }
    } else {
      analysis = {
        posture_score: 82,
        body_composition: { muscle_mass: 'good', body_fat: 'average' },
        measurements: { shoulders: 'aligned', spine: 'straight' },
        recommendations: ['Great posture! Work on core strength for improvement.']
      };
    }

    const xpAwarded = calculateXPReward('body');

    const { data: scanResult, error: insertError } = await supabase
      .from('scanner_results')
      .insert({
        user_id: userId,
        scan_type: 'body',
        image_url: imageDataUrl,
        analysis_data: analysis,
        xp_awarded: xpAwarded
      })
      .select()
      .single();

    if (insertError) throw insertError;

    const xpUpdate = await updateUserXP(userId, xpAwarded);

    res.json({
      message: 'Body scan completed successfully',
      scan: scanResult,
      analysis,
      xp_awarded: xpAwarded,
      ...xpUpdate
    });
  } catch (error) {
    console.error('Body scan error:', error);
    res.status(500).json({ error: 'Body scan failed' });
  }
});

router.post('/face', authenticateUser, upload.single('image'), async (req, res) => {
  try {
    const userId = req.user.id;
    const imageFile = req.file;

    if (!imageFile) {
      return res.status(400).json({ error: 'Image file is required' });
    }

    const base64Image = imageFile.buffer.toString('base64');
    const imageDataUrl = `data:${imageFile.mimetype};base64,${base64Image}`;

    let analysis = {
      skin_health_score: 0,
      skin_type: '',
      concerns: [],
      recommendations: []
    };

    if (process.env.OPENAI_API_KEY && process.env.OPENAI_API_KEY !== 'dummy-key') {
      try {
        const response = await openai.chat.completions.create({
          model: 'gpt-4o',
          messages: [
            {
              role: 'user',
              content: [
                {
                  type: 'text',
                  text: 'Analyze this facial image for skin health and provide: 1) Skin health score (0-100), 2) Skin type, 3) Visible concerns, 4) Beauty and skincare recommendations. Format as JSON.'
                },
                {
                  type: 'image_url',
                  image_url: { url: imageDataUrl }
                }
              ]
            }
          ],
          max_tokens: 500
        });

        const aiResponse = response.choices[0].message.content;
        try {
          analysis = JSON.parse(aiResponse);
        } catch {
          analysis.recommendations = [aiResponse];
        }
      } catch (aiError) {
        console.error('OpenAI API error:', aiError);
        analysis.recommendations = ['AI analysis unavailable. Using mock data.'];
      }
    } else {
      analysis = {
        skin_health_score: 85,
        skin_type: 'combination',
        concerns: ['minor dryness'],
        recommendations: ['Your skin looks healthy! Stay hydrated and use moisturizer daily.']
      };
    }

    const xpAwarded = calculateXPReward('face');

    const { data: scanResult, error: insertError } = await supabase
      .from('scanner_results')
      .insert({
        user_id: userId,
        scan_type: 'face',
        image_url: imageDataUrl,
        analysis_data: analysis,
        xp_awarded: xpAwarded
      })
      .select()
      .single();

    if (insertError) throw insertError;

    const xpUpdate = await updateUserXP(userId, xpAwarded);

    res.json({
      message: 'Face scan completed successfully',
      scan: scanResult,
      analysis,
      xp_awarded: xpAwarded,
      ...xpUpdate
    });
  } catch (error) {
    console.error('Face scan error:', error);
    res.status(500).json({ error: 'Face scan failed' });
  }
});

router.get('/history', authenticateUser, async (req, res) => {
  try {
    const userId = req.user.id;
    const { scan_type } = req.query;

    let query = supabase
      .from('scanner_results')
      .select('*')
      .eq('user_id', userId)
      .order('created_at', { ascending: false });

    if (scan_type) {
      query = query.eq('scan_type', scan_type);
    }

    const { data, error } = await query;

    if (error) throw error;

    res.json({ scans: data });
  } catch (error) {
    console.error('Scan history error:', error);
    res.status(500).json({ error: 'Failed to fetch scan history' });
  }
});

router.get('/stats', authenticateUser, async (req, res) => {
  try {
    const userId = req.user.id;

    const { data: scans, error } = await supabase
      .from('scanner_results')
      .select('scan_type, xp_awarded, created_at')
      .eq('user_id', userId);

    if (error) throw error;

    const stats = {
      total_scans: scans.length,
      scans_by_type: {
        food: scans.filter(s => s.scan_type === 'food').length,
        body: scans.filter(s => s.scan_type === 'body').length,
        face: scans.filter(s => s.scan_type === 'face').length
      },
      total_xp_earned: scans.reduce((sum, s) => sum + s.xp_awarded, 0),
      recent_scans: scans.slice(0, 5)
    };

    res.json(stats);
  } catch (error) {
    console.error('Scan stats error:', error);
    res.status(500).json({ error: 'Failed to fetch scan stats' });
  }
});

export default router;
